src/atldld/sync.py

```diff
@@ -36,6 +36,113 @@
 )
 
 
+def xy_to_pir(
+    coords_img: np.ndarray,
+    affine_2d: np.ndarray,
+    affine_3d: np.ndarray,
+) -> np.ndarray:
+    """Transform coordinates from the image space to the reference space.
+
+    Parameters
+    ----------
+    coords_img
+        Array of shape `(3, N)` where the first axis contains the
+        `x`, `y` and `section_number * section_thickness`. Note that
+        both that the `section_number` (image specific) and
+        `section_thickness` (dataset specific) can be retrieved from the
+        Allen Brain API.
+    affine_2d
+        Matrix of shape `(2, 3)` representing a 2D affine transformation. It
+        can be retrieved from the section image metadata via the
+        Allen Brain API. More specifically, it is stored under the
+        `tvr_**` entries.
+    affine_3d
+        Matrix of shape `(3, 4)` representing a 3D affine transformation. It
+        can be retrieved from the dataset metadata via the Allen Brain API.
+        More specifically, it is stored under the `tsv_**` entries.
+
+    Returns
+    -------
+    coords_ref : np.ndarray
+        Array of shape `(3, N)` where the first axis contains the `p`, `i`,
+        `r` coordinates.
+    """
+    dtype = np.float32
+    n_coords = coords_img.shape[1]
+
+    coords_img_ = np.concatenate([coords_img, np.ones((1, n_coords), dtype=dtype)])
+
+    affine_2d_ = np.array(
+        [
+            [affine_2d[0, 0], affine_2d[0, 1], 0, affine_2d[0, 2]],
+            [affine_2d[1, 0], affine_2d[1, 1], 0, affine_2d[1, 2]],
+            [0, 0, 1, 0],
+            [0, 0, 0, 1],
+        ],
+        dtype=dtype,
+    )
+    affine_3d_ = affine_3d.astype(dtype)
+    coords_ref = (affine_3d_ @ affine_2d_) @ coords_img_
+
+    return coords_ref
+
+
+def pir_to_xy(
+    coords_ref: np.ndarray,
+    affine_2d: np.ndarray,
+    affine_3d: np.ndarray,
+) -> np.ndarray:
+    """Transform coordinates from the reference space to the image space.
+
+    Parameters
+    ----------
+    coords_ref
+        Array of shape `(3, N)` where the first axis contains the
+        `p`, `i` and `r`.
+    affine_2d
+        Matrix of shape `(2, 3)` representing a 2D affine transformation. It
+        can be retrieved from the section image metadata via the
+        Allen Brain API. More specifically, it is stored under the
+        `trv_**` entries.
+    affine_3d
+        Matrix of shape `(3, 4)` representing a 3D affine transformation. It
+        can be retrieved from the dataset metadata via the Allen Brain API.
+        More specifically, it is stored under the `tvs_**` entries.
+
+    Returns
+    -------
+    coords_img : np.ndarray
+        Array of shape `(3, N)` where the first axis contains the `x`, `y`,
+        `section_number * section_thickness`. Note that the `section_number`
+        can be an arbitrary float that is most likely not equal to a
+        `section_number` of any section image.
```

Comment:
So I think this `section_number` is just a counting index of the sections as they were made in the lab. Since sections of one specimen were used for multiple stainings there are gaps in this index.

Just for a test I got all datasets for a given speciment ID and listed all section numbers in a sorted way:
```python
#!/usr/bin/env python
import pandas as pd

from atldld.requests import RMAParameters, rma_all

specimen_id = 702765

params = RMAParameters(
    "SectionDataSet",
    criteria={"specimen_id": specimen_id},
    include=["section_images", "genes"],
)
dataset_metas = rma_all(params)

print(f"Found {len(dataset_metas)} datasets for specimen ID {specimen_id}")
for ds_meta in sorted(dataset_metas, key=lambda m: m["id"]):
    print(
        f'Dataset ID {ds_meta["id"]:5d} : '
        f'{", ".join(gene["acronym"] for gene in ds_meta["genes"])}'
    )

df_sections = pd.DataFrame(
    columns=("Section Number", "Image ID", "Dataset ID", "Genes"),
    data=[
        [
            img_meta["section_number"],
            img_meta["id"],
            img_meta["data_set_id"],
            ", ".join(gene["acronym"] for gene in ds_meta["genes"]),
        ]
        for ds_meta in dataset_metas
        for img_meta in ds_meta["section_images"]
    ],
)
print()
print("All section images sorted by section number")
print(df_sections.sort_values("Section Number", ignore_index=True))
```
Output:
```
Found 8 datasets for specimen ID 702765
Dataset ID    33 : Anln
Dataset ID    67 : Actr2
Dataset ID   479 : Gad1
Dataset ID   659 : Mapre2
Dataset ID   715 : Npdc1
Dataset ID   760 : Olfm1
Dataset ID  1211 : Gphn
Dataset ID  1261 : Pfn2

All section images sorted by section number
     Section Number   Image ID  Dataset ID   Genes
0                 0  101349752         715   Npdc1
1                 1  101349753          33    Anln
2                 2  101349754         659  Mapre2
3                 3  101349755        1211    Gphn
4                 4  101349756         479    Gad1
..              ...        ...         ...     ...
440             443  101350195        1211    Gphn
441             444  101350196         479    Gad1
442             445  101350197          67   Actr2
443             446  101350198         760   Olfm1
444             447  101350199        1261    Pfn2

[445 rows x 4 columns]
```
Observations:
- Section numbers go from 0 to 447
- Seems like 3 sections are missing, maybe failed slices?
- Image IDs are consecutive as well.
- Genes are interspered.

Reply:
Interesting analysis! BTW this docstring is for the `pir_to_xy` and the reason why I say that it is a **theoretical section number** is because the chances are that if you take a random `(p, i, r)`  then the section number after the mapping will not be an integer. However, AFAIK the actual reference-to-image API actually gives the user the closest section image to it ([link](http://help.brain-map.org/display/api/Image-to-Image+Synchronization#Image-to-ImageSynchronization-Reference-To-Image))

Reply:
Missing sections:
```
[163] : 162 (101349914, Mapre2) -> 164 (101349916, Gad1)
[171] : 170 (101349922, Mapre2) -> 172 (101349924, Gad1)
[283] : 282 (101350034, Mapre2) -> 284 (101350036, Gad1)
```
Seems like something went wrong with 3 `Gphn` stainings.

Reply:
Really interesting analysis @Stannislav :)